File: src/GameLogic.py
import time
from Player import Player
from constants import *
from board_validations import *
import constants
from History import History
from Screen import Screen


class GameLogic:
    def __init__(self, game_config: dict[str, str | bool]):
        self.board = [[0 for _ in range(constants.COLS)]
                      for _ in range(constants.ROWS)]

        self.player1 = Player(1)                        # Player 1 is black
        self.player2 = Player(-1, game_config["ai"])    # Player 2 is white
        self.current_player = self.player1
        self.inactive_player = self.player2

        self.current_player.color_start_time = time.time()
        self.save_history = game_config["save_history"]
        if self.save_history:
            self.game_history = History()

        # When game launches, draw the board
        self.screen = Screen()
        self.screen.draw_board()
        self.screen.draw_player_info(
            self.current_player.num, self.player1, self.player2)

    def handle_capture(self, stones_to_delete) -> bool:
        for stone in stones_to_delete:
            self.board[stone[1]][stone[0]] = 0
        self.screen.draw_board()
        self.screen.draw_all_stones(self.board)
        if self.current_player.num == 1:
            if self.player1.handle_taken_stones():
                return True
        else:
            if self.player2.handle_taken_stones():
                return True
        return False

    def check_capture(self, cell):
        capture_check_result = check_capture(
            self.board, cell, self.current_player.num)
        if capture_check_result:
            player = self.current_player.num
            print(f"{T_CYAN}{player}'s stones were captured!{T_GRAY}")
            if self.save_history:
                self.game_history.add_capture(
                    self.current_player, self.inactive_player)
            if self.handle_capture(capture_check_result):
                self.draw_win_screen()
                return False
        return True

    def check_game_continue(self, cell):
        if check_win(self.board, cell, self.current_player.num):
            self.screen.draw_win_screen(self.current_player.name)
            if self.save_history:
                self.game_history.add_win(self.current_player.name)
            print(f"{T_GREEN}{self.current_player} player wins!!")
            return False

        # Check if the board is full
        if check_board_full(self.board):
            print(f"{T_YELLOW}It's a tie!{T_GRAY}")
            return False

        return True

    def apply_move(self, cell):
        col, row = cell
        self.board[row][col] = self.current_player.num

        if self.save_history:
            self.game_history.add_move(cell, self.current_player.num)

        self.check_capture(cell)

        if not self.check_game_continue(cell):
            return False

        self.screen.draw_player_info(
            self.current_player.num, self.player1, self.player2)
        self.screen.draw_stone(
            cell, BLACK if self.current_player.num == 1 else WHITE)

        # Change active player
        temp = self.current_player
        self.current_player = self.inactive_player
        self.inactive_player = temp

        return True

    def handle_turn(self, cell):
        # Checks (to be implemented)

        return self.apply_move(cell)

    """ def handle_ai_turn(self):
        result = self.player2.new_ai_move(self.board)
        print(f"{T_YELLOW}AI move: {result[0]}, {result[1]}{T_GRAY}")

        self.player2.timer = result[3]

        self.current_player.num= 1
        self.current_player.color_start_time = time.time()

        if result[0] == -1 and result[1] == -1:
            return False

        self.board[result[0]][result[1]] = -1
        self.draw_stone((result[1], result[0]), WHITE)

    def handle_turn(self, cell) -> bool:
        # x, y = cell
        col, row = cell

        # Calculate time taken for this move
        current_time = time.time()
        time_taken = current_time - self.current_player.color_start_time

        # Update the appropriate player's timer
        if self.current_player.num== 1:
            self.player1.timer = time_taken
        else:
            if not self.player2.ai:
                self.player2.timer = time_taken

        # Check if the cell is already occupied
        if self.board[row][col] != 0:
            print(f"{T_BLUE}Can't place on an occupied tile! Invalid move!{T_GRAY}")
            return True

        # Check for double three
        if check_double_three(self.board, cell, self.current_player.num):
            print(f"{T_PURPLE}Double three detected! Invalid move!{T_GRAY}")
            return True

        # Check if new position moves into capture
        if check_move_into_capture(self.board, cell, self.current_player.num):
            print(f"{T_RED}Can't move into capture! Invalid move!{T_GRAY}")
            return True

        self.board[row][col] = self.current_player.num

        # Check if there is a capture
        capture_check_result = check_capture(self.board, cell, self.current_player.num)
        if capture_check_result:
            player = "Black" if self.current_player.num== -1 else "White"
            print(f"{T_CYAN}{player}'s stones were captured!{T_GRAY}")
            if self.save_history:
                self.game_history.add_capture(player)
            if self.handle_capture(capture_check_result):
                self.draw_win_screen()
                return False

        # Check if it's a winning move
        if check_win(self.board, cell, self.current_player.num):
            self.draw_win_screen()
            return False

        # Draw the stone
        self.draw_stone(cell, BLACK if self.current_player.num== 1 else WHITE)
        self.current_player.num= -self.current_player.num

        self.current_player.color_start_time = time.time()

        # Check if the board is full
        if check_board_full(self.board):
            self.draw_tie_screen()
            return False

        if self.current_player.num== -1 and self.player2.ai:
            self.handle_ai_turn()

        self.draw_player_info()

        return True
 """

File: c_gomoku/src/classes/GameLogic.cpp
/* ************************************************************************** */
/*                                                                            */
/*                                                        :::      ::::::::   */
/*   GameLogic.cpp                                      :+:      :+:    :+:   */
/*                                                    +:+ +:+         +:+     */
/*   By: alaparic <alaparic@student.42.fr>          +#+  +:+       +#+        */
/*                                                +#+#+#+#+#+   +#+           */
/*   Created: 2025/05/12 22:06:26 by alaparic          #+#    #+#             */
/*   Updated: 2025/06/20 13:27:08 by alaparic         ###   ########.fr       */
/*                                                                            */
/* ************************************************************************** */

#include <iostream>
#include "GameLogic.hpp"
#include <unistd.h>

GameLogic::GameLogic() {}

GameLogic::GameLogic(s_game_config config)
{
	this->board_size = config.board_size;
	this->save_history = config.save_history;
	this->subject_mode = (config.game_mode == "subject");
	this->gameMode = config.game_mode;

	// Initialize players
	this->player1 = Player(BLACK_STONE, false);
	this->player2 = Player(WHITE_STONE, config.is_ai);
	this->currentPlayer = &this->player1;
	this->inactivePlayer = &this->player2;

	this->currentPlayer->startTimer();

	this->board = new Board(config.board_size, config);
	this->screen = new Screen(this->board_size);
	this->history = new History();
}

GameLogic::~GameLogic()
{
	delete this->screen;
	delete this->board;
	delete this->history;
}

/* Getter */

History *GameLogic::getHistory() const
{
	return this->history;
}

/* Mouse Callback */
void GameLogic::mouseButtonCallback(mouse_key_t button, action_t action, modifier_key_t mods, void *param)
{
	(void)mods;

	GameLogic *game = static_cast<GameLogic *>(param);
	if (action == MLX_PRESS && button == MLX_MOUSE_BUTTON_LEFT)
	{
		int xpos, ypos;
		mlx_get_mouse_pos(game->screen->getMLX(), &xpos, &ypos);

		// Calculate the closest intersection
		int cell_size = (SCREEN_SIZE - 2 * MARGIN) / (game->board_size - 1);
		int col = std::round((xpos - MARGIN) / static_cast<double>(cell_size));
		int row = std::round((ypos - MARGIN) / static_cast<double>(cell_size));

		// Ensure the position is within bounds
		if (col < 0 || col >= game->board_size || row < 0 || row >= game->board_size)
			return;

		std::pair<int, int> cell = {col, row};
		int continueGame = game->handleTurn(cell);
		if (!continueGame)
			mlx_close_window(game->screen->getMLX());
		else if (game->player2.isAI() && game->currentPlayer == &game->player2)
		{
			std::pair<int, int> aiMove = game->player2.getAIMove(game->board);
			if (!game->handleTurn(aiMove))
				mlx_close_window(game->screen->getMLX());
		}
	}
}

/* Methods */

bool GameLogic::checkIllegalMove(const std::pair<int, int> &cell)
{
	int col = cell.first;
	int row = cell.second;

	// Check if the move is out of bounds
	if (!this->board->inBounds(col, row))
		return true;

	// Check if the cell is already occupied
	if (checkOccupiedCell(*this->board, col, row))
		return true;

	if (this->gameMode != "standard")
	{
		// Check if the move creates a double three
		if (checkDoubleThree(*this->board, cell, this->currentPlayer->getColor()))
		{
			std::cout << T_PURPLE << "Illegal move: Double three detected!" << T_GRAY << std::endl;
			return true;
		}

		// Check if the move is into a capture
		if (checkMoveIntoCapture(*this->board, cell, this->currentPlayer->getColor()))
		{
			std::cout << T_PURPLE << "Illegal move: Cannot move into capture!" << T_GRAY << std::endl;
			return true;
		}
	}

	return false;
}

bool GameLogic::checkGameEnd(const std::pair<int, int> &lastMove)
{
	int x = lastMove.first;
	int y = lastMove.second;
	if (this->board->checkWin(x, y, this->currentPlayer->getColor()))
	{
		std::cout << T_WHITE << "-----------------------------------" << std::endl;
		std::cout << T_YELLOW << "Five in a row!" << std::endl;
		std::cout << T_GREEN << this->currentPlayer->getName() << " Player wins!" << T_GRAY << std::endl;
		if (this->save_history)
			this->history->addWin(this->currentPlayer);
		return true;
	}
	else if (this->currentPlayer->getTakenStones() >= 10 && this->gameMode == "subject")
	{
		std::cout << T_WHITE << "-----------------------------------" << std::endl;
		std::cout << T_YELLOW << "The game is over due to " << this->currentPlayer->getTakenStones() << " stones taken!" << std::endl;
		std::cout << T_GREEN << this->currentPlayer->getName() << " Player wins!" << T_GRAY << std::endl;
		if (this->save_history)
			this->history->addWin(this->currentPlayer);
		return true;
	}
	else if (this->board->isFull())
	{
		std::cout << T_WHITE << "-----------------------------------" << std::endl;
		std::cout << T_YELLOW << "The game is a draw!" << T_GRAY << std::endl;
		if (this->save_history)
			this->history->addTie();
		return true;
	}
	return false;
}

bool GameLogic::applyMove(const std::pair<int, int> &cell)
{
	int col = cell.first;
	int row = cell.second;

	// ! Temp
	std::cout << T_CYAN << "Board evaluation for " << this->currentPlayer->getName() << " Player:" << std::endl;
	std::cout << staticBoardEvaluation(this->board, this->currentPlayer->getColor()) << T_GRAY << std::endl;

	// Place stone on board
	this->board->set(col, row, this->currentPlayer->getColor());

	// Draw the stone on screen
	this->screen->drawStone(col, row, this->currentPlayer->getColor() == BLACK_STONE ? BLACK : WHITE);

	this->currentPlayer->stopTimer();

	// Add move to history
	if (this->save_history)
		this->history->addMove(cell, this->currentPlayer);

	// Check for capture
	if (this->gameMode != "standard")
	{
		std::vector<std::pair<int, int>> capturedStones = checkCapture(*this->board, cell, this->currentPlayer->getColor());
		if (!capturedStones.empty())
		{
			std::cout << T_YELLOW << "Capture detected!" << T_GRAY << std::endl;
			this->handleCapture(capturedStones);

			// Add capture to history
			if (this->save_history)
				this->history->addCapture(this->currentPlayer, this->inactivePlayer);
		}
	}

	// Check for end game conditions
	if (checkGameEnd(cell))
		return false;

	// Change active player
	Player *temp = this->currentPlayer;
	this->currentPlayer = this->inactivePlayer;
	this->inactivePlayer = temp;

	this->currentPlayer->startTimer();

	// Update player info display
	this->screen->drawPlayerInfo(&this->player1, &this->player2, this->currentPlayer);

	return true;
}

bool GameLogic::handleTurn(const std::pair<int, int> &cell)
{

	if (this->checkIllegalMove(cell))
		return true;

	return this->applyMove(cell);
}

void GameLogic::handleCapture(const std::vector<std::pair<int, int>> &takenStones)
{
	this->currentPlayer->addTakenStones(takenStones.size());
	for (const std::pair<int, int> &stone : takenStones)
	{
		int col = stone.first;
		int row = stone.second;

		if (this->gameMode == "subject")
			this->board->set(col, row, EMPTY);
		else if (this->gameMode == "reversi")
			this->board->set(col, row, this->currentPlayer->getColor());
	}
	this->screen->clearArea(0, 0, SCREEN_SIZE, SCREEN_SIZE, BLACK);
	this->screen->drawBoard();
	this->screen->drawAllStones(this->board);
}

void GameLogic::startGame()
{
	// Set up the mouse callback
	mlx_mouse_hook(this->screen->getMLX(), &GameLogic::mouseButtonCallback, this);

	// Draw the initial board
	this->screen->drawBoard();

	// Draw player info
	this->screen->drawPlayerInfo(&this->player1, &this->player2, this->currentPlayer);

	// Start the main loop
	mlx_loop(this->screen->getMLX());
}
